sdk/core/Azure.Core/src/Shared/OperationInternalOfT.cs

```diff
@@ -153,44 +149,26 @@ private set
         /// <exception cref="RequestFailedException">Thrown if there's been any issues during the connection, or if the operation has completed with failures.</exception>
         public async ValueTask<Response<T>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
         {
-            while (true)
-            {
-                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
-
-                if (HasCompleted)
-                {
-                    return Response.FromValue(Value, response);
-                }
-
-                GetServerDelay(response, pollingInterval);
-            }
+            var rawResponse = await WaitForCompletionResponseAsync(pollingInterval, cancellationToken).ConfigureAwait(false);
+            return Response.FromValue(Value, rawResponse);
         }
 
         protected override async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
         {
             OperationState<T> state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
-
-            RawResponse = state.RawResponse;
-
-            if (state.HasCompleted)
+            try
             {
-                if (state.HasSucceeded)
+                if (state.HasCompleted && state.HasSucceeded)
                 {
-                    Value = state.Value;
-                    HasCompleted = true;
-                }
-                else
-                {
-                    _operationFailedException = state.OperationFailedException ?? (async
-                        ? await _diagnostics.CreateRequestFailedExceptionAsync(state.RawResponse).ConfigureAwait(false)
-                        : _diagnostics.CreateRequestFailedException(state.RawResponse));
-                    HasCompleted = true;
-
-                    throw _operationFailedException;
+                    Value = state.Value!;
                 }
+                return await ApplyStateAsync(async, state.RawResponse, state.HasCompleted, state.HasSucceeded, state.OperationFailedException).ConfigureAwait(false);
+            }
+            catch (RequestFailedException ex)
+            {
+                _operationFailedException = ex;
```

Comment:
Is it worth storing the exceptin in the base class even though the OperationInternal doesn't use it? Might simpligy this type a bit.

Reply:
Ah, you do store it there. So should we make it protected?

Reply:
Fixed